### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] Counts type parameter references in body irrelevant to the signature

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#9735]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.tsx&code=GYVwdgxgLglg9mABMOcA8AVAfACgPoBciGAlIgN4BQiiECAzlIgB5EaIC8iYIANrwEIA3NVoMmATzadufQSIC%2BlIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function foo<T>(_: T) {
  const x: T = null!;
  const y: T = null!;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-parameters": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`T` should be marked as an unnecessary type parameter. It's only used once in the type signature.

### Actual Result

No report.

### Additional Info

Repro case courtesy of @That-Guy977's https://github.com/typescript-eslint/typescript-eslint/issues/9664#issuecomment-2256193242, thanks!

💖

**Comment by ST-DDT** [comment]
Could you please explain why type parameter usages in the implementation should be ignored?
Neither the issue description nor the closing PR explains this.

**Comment by Josh-Cena** [comment]
Type parameters in the body is type checked using the base constraint of the parameter, so `: T` should be equivalent to `: unknown`. Type parameters are not realized until the function gets called, by which time the function body has been type checked already.

**Comment by That-Guy977** [comment]
> Type parameters in the body is type checked using the base constraint of the parameter, so `: T` should be equivalent to `: unknown`.

Aren't they non-equivalent when assigning to `T`, acting more like `never`?

```ts
function f<T>() {
  const x: T = ""
//      ^
// Type 'string' is not assignable to type 'T'.
//   'T' could be instantiated with an arbitrary type which could be unrelated to 'string'.
}
```

**Comment by Josh-Cena** [comment]
In a way, they are the widest possible type when assigned from and the narrowest possible type when assigned to. I realized how tricky it is so hopefully my second paragraph is a better phrasing of what I meant.

**Comment by bradzacher** [comment]
Putting it another way:

A type parameter `T` bounded by `extends U` acts like a transparent[^1] nominal type.
It is only assignable to `any`, `unknown`, `T` or `U`
It only accepts `any`, `never` or `T`

Note: for unbounded type parameter (eg `<T>`) is the same as `<T extends unknown>` and thus `U = unknown` above.

[^1]: Side note that the only difference between a transparent nominal and an opaque nominal type is that an opaque nominal type disallows the assignment to `U`. [This is why I like flow cos it lets you declare both](https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZlduuSMTBJC-Mgz4sHwNSBos6A6u9w1mbGqJUoBaCRHEzrcDEgBrbAk62mPhXFxJ91v7x+fRHEp3gxCKBgNpUU7nAF5KBQA7tEhUJ4VJqmDA+SEgIEwsBwhHHCCMTZQbjHOoYCrooHwmDYY4Ie5HSgkY5kCQM7ZXKH5E7MdJwMgAgC+-EgGiSa1wVGwAAJ8RgAI4yKVnYxSgAq+FVNDsjhYPmQUvspigCAA3AAdKDiyVS5FKzWopiSvVqjVax26iTmqAWq0sGWOBXS5XS9UAeQDMi9vultok4flMn1YYj2C9Fts9il8gAjPrDRUEFKALxK+OBr0ZiRZgBMeakBeLdrdOosaZ8ZcVJYA5F2vRJ7dqnVXu72LbkQA0TCQ4NAkg0AAxWasAFmXVnnIAFQA)
